sys/psa_crypto/psa_persistent_storage/psa_crypto_cbor_encoder.c

```diff
@@ -0,0 +1,625 @@
+        default:
+            return "Error value not recognized";
+    }
+}
+
+/**
+ * @brief   Convert nanocbor error to PSA Crypto status value
+ *
+ * @param   res Nanocbor error value
+ * @return  psa_status_t
+ */
+static psa_status_t nanocbor_error_to_psa_status(int res)
+{
+    switch (res) {
+        case NANOCBOR_ERR_END:
+            return PSA_ERROR_BUFFER_TOO_SMALL;
+        case NANOCBOR_NOT_FOUND:
+            return PSA_ERROR_DOES_NOT_EXIST;
+        default:
+            return PSA_ERROR_GENERIC_ERROR;
+    }
+}
+
+/**
+ * @brief   Encode key data stored in a PSA key slot in CBOR.
+ *
+ * @param   enc             Encode object of type @c nanocbor_encoder_t
+ * @param   slot            Key slot containing the key to encode
+ * @param   is_public_key   True if key to encode is asymmetric public key,
+ *                          false if it's a single key or asymmetric private key
+ *
+ * @return  psa_status_t
+ */
+static psa_status_t psa_encode_key_data(nanocbor_encoder_t *enc, uint8_t *key_data,
+                                        size_t key_data_len)
+{
+    int res;
+
+    res = nanocbor_fmt_uint(enc, key_data_len);
+    if (res < 0) {
+        goto error;
+    }
+
+    res = nanocbor_fmt_array(enc, key_data_len);
+    if (res < 0) {
+        goto error;
+    }
+
+    for (size_t i = 0; i < key_data_len; i++) {
+        res = nanocbor_fmt_uint(enc, key_data[i]);
+        if (res < 0) {
+            goto error;
+        }
+    }
+
+    return PSA_SUCCESS;
+
+error:
+    DEBUG("[persistent storage] error while encoding key data: %d = %s, ", res,
+                                                nanocbor_error_to_string(res));
+    return nanocbor_error_to_psa_status(res);
+}
+
+#if PSA_PROTECTED_KEY_COUNT
+/**
+ * @brief   Encode contents of a PSA key slot containing a protected key
+ *
+ * @param   enc     Active nanocbor encoder
+ * @param   slot    Pointer to key slot to encode
+ * @return  psa_status_t
+ */
+static psa_status_t psa_encode_protected_key_slot(nanocbor_encoder_t *enc, psa_key_slot_t *slot)
+{
+    int res;
+    int key_array_size = 1;
+    psa_key_slot_number_t *slot_no = psa_key_slot_get_slot_number(slot);
+
+#if IS_USED(MODULE_PSA_ASYMMETRIC)
+    if (PSA_KEY_TYPE_IS_KEY_PAIR(slot->attr.type)) {
+        key_array_size = 2;
+    }
+#endif /* MODULE_PSA_ASYMMETRIC */
+
+    res = nanocbor_fmt_array(enc, key_array_size);
+    if (res < 0) {
+        goto error;
+    }
+
+    res = nanocbor_fmt_uint(enc, *slot_no);
+    if (res < 0) {
+        goto error;
+    }
+
+#if IS_USED(MODULE_PSA_ASYMMETRIC)
+    if (PSA_KEY_TYPE_IS_KEY_PAIR(slot->attr.type)) {
+        uint8_t *pubkey_data;
+        size_t *pubkey_data_len;
+        psa_get_public_key_data_from_key_slot(slot, &pubkey_data, &pubkey_data_len);
+
+        res = nanocbor_fmt_array(enc, 2);
+        if (res < 0) {
+            goto error;
+        }
+
+        psa_status_t status = psa_encode_key_data(enc, pubkey_data, *pubkey_data_len);
```

Comment:
what is the second array element here? If they are the two fields encoded in `psa_encode_key_data`, the `nanocbor_fmt_array` should be inside that function, I'd argue.

Reply:
Using bytestrings instead of arrays reduces the amount of necessary code, so I removed the `psa_encode_data` function completely.